internal/testrunner/runners/system/runner.go

```diff
@@ -220,6 +221,16 @@ func (r *runner) hasNumDocs(
 				multiErr = append(multiErr, errs...)
 				continue
 			}
+
+			// Generate sample event from first doc
```

Comment:
I do love simplicity of this solution :)

I'm only concerned about the fact that it's in `hasNumDocs`, which shouldn't modify something externally. Keep in mind that the function is actually called twice (on clear and on check). Will the file be written twice?

I'm wondering if we can replace `hasNumDocs` with `areDocsAvailable`, and define a set of callbacks for:
- minDocsNumFunc
- verifyDocumentFunc(data Document) error
- onFirstDocumentFunc(data Document) error - which can generate sample docs.

I'm happy to discuss it as it's actually a refactoring task.

Reply:
Ah, I didn't even realize the name of the function in which I added this code. I also didn't realize this function was being called twice. Thanks for pointing out these two things.

I completely agree that we need to refactor this code. Let me take a stab at it and have you review it.

Reply:
I did some refactoring in 3b0c4c3. Let me know what you think.

Reply:
I did some refactoring in 3b0c4c3.  Broadly speaking, the changes I made were:
- renamed `hasNumDocs` to `checkNumDocs`
- `checkNumDocs` now returns the number of documents + the documents themselves + any error
- don't perform field validation when `checkNumDocs` is called for clearing out old docs in the data stream
- don't write `sample_event.json` when `checkNumDocs` is called for clearing out old docs in the data stream

Let me know what you think.

Reply:
Hold on, I thought of a way to simplify things further.

Reply:
Okay, some more refactoring done in 33163be.